File: lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.parent = None
"""

class Solution:
    def lowestCommonAncestor(self, p: 'Node', q: 'Node') -> 'Node':
        
        seen = set()
        
        while p:
            seen.add(p)
            p = p.parent

        while q:
            if q in seen:
                return q
            q = q.parent
                        
        return None
            
        
"""
Cases:
- both have equal distance to parent
    - parent, parent, parent, if equal, return...
- p has longer path than q
    - go up p
    - go up q, if q in seen_p, return!
- q has longer path than p

"""        

File: lowest-common-ancestor-of-a-binary-tree-iii/lowest-common-ancestor-of-a-binary-tree-iii.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var left: Node?
 *     public var right: Node?
 *     public var parent: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *         self.parent = nil
 *     }
 * }
 */

class Solution {
    func lowestCommonAncestor(_ p: Node?,_ q: Node?) -> Node? {
        
        var seen: Set<Int> = []
        var a:Node? = p
        var b:Node? = q
        
        while a != nil {
            seen.insert(a?.val ?? 0)
            a = a?.parent
        }
        
        while b != nil {
            if seen.contains(b?.val ?? 0) {
                return b
            }
            b = b?.parent
        }
        
        return a
    }
}